---
title: "Simulation 2 Documentation"
author: "Hillary Heiling"
date: "April 27, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Overview

This document performs the steps, outlined in "Simulation Step Overview.Rmd", used to set-up and perform simulations to test the package `IsoDeconvMM`. 

Unlike "Simulation 1 Documentation.Rmd" and "Simulation 2 Documentation.Rmd", the exon set counts are modeled with a Dirichlet-Multinomial distriubtion (the other simulations utilize a Dirichlet-Negative Binomial model).

Like the Fit 2 in both "Simulation 1 Documentation.Rmd" and "Simulation 2 Documentation.Rmd", Blueprint alpha results will be used in the determination of exon set counts for the genes/clusters specified to have differential isoform usage. Unlike Fit 2 of these past simulation results, these alpha results can be used directly without any additional scaling. 

Details of simulation:

A total of 5,172 genes were selected for the simulation (see "Gene Samples and EffLen Design Matrix.Rmd" for documentation of how these genes were selected). Of these approx. 5,000 genes, 1,000 were chosen to be 'genes of interest', where various subsets of these 1,000 genes will be used in the `IsoDeconvMM` fitting procedure (see "Gene Samples and EffLen Design Matrix.Rmd" for documentation of how these 1,000 genes were selected). 

100 of the 1,000 'genes of interest' will be selected for differential expression, and another 100 of the 1,000 'genes of interest' (non-overlapping with the genes chosen for differential expression) will be selected for differential isoform usage.

Initial gene counts for all 5,172 genes will be determined by (a) randomly selecting a probability vector from the Dirichlet distribution fit to the 5,172 total genes based on UCLA data (documentation in "Gene Samples and EffLen Design Matrix.Rmd") and (b) simulating a specified total read count number from a multinomial distribution with the probability vector from (a). Genes specified for differential expression will have their gene counts modified in one of the cell types.

Exon set counts for the 1,000 genes of interest will be calculated based on the procedure outlined in "Simulation Step Overview.Rmd". Exon set counts for the other 4,172 genes will be arbitrarily distributed approx. uniformly over all of the possible single exon sets for the genes. For the remaining 4,172 genes that are not the 1,000 'genes of interest', these genes primary function are to contribute to a realistic distribution of gene counts. Therefore, we only need to calculate a realistic gene count for each of these genes (based on the Dirichlet model fit to the UCLA data); we don't necessarily care about a 'realistic' distribution of counts across isoforms and specific exon sets because these gene clusters will not be used in the `IsoDeconvMM` fit algorithm.

Once the exon set counts are determined for all pure cell type reference samples, exon set counts for mixture samples will be created from a subset of these reference sample results (the other samples not used for mixture sample creation will be used in the `IsoDeconvMM` model fit).

## Load Materials

Load the materials created in previous documentation that will be needed for the simulation

* nTE_filtered: data.frame containing geneId and clustID information for the 1,000 'genes of interest' (to be used in the `IsoDeconvMM` fit algorithm) sutsetted from the 5,172 total genes selected for simulation. Procedure of creating nTE_filtered from original nTE object documented in "Gene Samples and EffLen Design Matrix.Rmd"

* other_geneInfo: list containing nTE_other and exon_sets_other. nTE_other is a data.frame containing geneId and clustID information for the other (5,172 - 1,000 = 4,172) genes selected for the simulation but not of particular interest for the `IsoDeconvMM` fit algorithm. The exon_sets_other is a character vector of sigular exon sets for these 4,172 genes. Procedure to create these items documented in "Gene Samples and EffLen Design Matrix.Rmd"

* X_info: list containing effective length design matrix and exon set information for each of the 1,000 clusters of interest. Procedure to create this object documented in "Gene Samples and EffLen Design Matrix.Rmd"

* genes_alpha: alpha values for the Dirichlet distribution needed for the "Total Reads to Gene Parameters" section outlined in "Simulation Step Overview.Rmd". Fitting of Dirichlet model and calculation of alpha values documented in "Gene Samples and EffLen Design Matris.Rmd"

* isoAll: known isoforms object for human data. When loaded, this object is a list where each component is a binary matrix that specifies a set of possible isoforms (e.g., isoforms from annotations). Specifically, it is a binary matrix of k rows and m columns, where k is the number of 
non-overlapping exons and m is the number of isoforms. isoforms[i,j]=1 indicates that the i-th exon belongs to the j-th isoform. For example, the following matrix indicates the three isoforms for one gene ENSMUSG00000000003:

```
     ENSMUST00000000003 ENSMUST00000166366 ENSMUST00000114041
[1,]                  1                  1                  1
[2,]                  1                  1                  1
[3,]                  1                  1                  1
[4,]                  1                  1                  0
[5,]                  1                  1                  1
[6,]                  1                  1                  1
[7,]                  1                  1                  1
[8,]                  1                  0                  0
```

Instructions for creating such an RData object can be found in the Create_BED_knownIsoforms_Files directory in the GitHub repo hheiling/deconvolution, <https://github.com/hheiling/deconvolution>

Load materials

```{r}
# nTE_filtered
load("Human_Materials/Homo_sapiens.GRCh37.66.nTE.filtered.RData")

# nTE
load("Human_Materials/Homo_sapiens.GRCh37.66.nTE.RData")

# X_info
load("Human_Materials/Effective Length Design Matrices.RData")

# genes_alpha
load("Human_Materials/Dirichlet_UCLA_gene_alpha.RData")

# isoAll
load("Human_Materials/Homo_sapiens.GRCh37.66.nonoverlap.exon.knownIsoforms.RData")

# other_geneInfo
load("Human_Materials/Homo_sapiens.GRCh37.66.other.geneInfo.RData")

# alpha_collect
# load("Blueprint_Alpha_Collection.RData") # 7A series alpha results
# load("Blueprint_Alpha_Collection_9A.RData") # 9A series alpha results
# alpha_collect11A
load("Blueprint_Alpha_Collection_11A.RData") # 11A series alpha results
```

Needed libraries

```{r}
library(DirichletReg)
library(stringr)
library(Rcpp)
library(RcppArmadillo)

```

Source relevent functions needed for simulations in "Simulation_Ideas/geneModel_code" folder

```{r}
source("geneModel_code/sim_functions.R")
```

```{r, eval=FALSE}
sourceCpp(file = "geneModel_code/fragLens_dist.cpp")
```

Generate random seeds to be used in the simulation

```{r}
set.seed(2020)
seeds = sample(1000:9999, size = 50, replace = F)
seeds
```

## Calculate Relevant Parameters and Output for Cell Types 1, 2, and 3

Specifying n = number pure cell reference files per cell type and total_cts = total gene counts for each sample

Details:

* Want 3 cell types

* Want 20 pure cell reference files per cell type

* Assume total read counts is distributed normally with mean 7 million and sd 1 million

```{r}
# Number pure cell reference files to create
n = 20
# Number cell types
J = 3

# Total counts for each sample
set.seed(seeds[1])
total_cts = matrix(round(rnorm(n = n*J, mean = 7*10^6, sd = 10^6)), nrow = n, ncol = J)

colnames(total_cts) = 1:J
rownames(total_cts) = str_c("ref_",1:n)

```

Calculate initial gene counts across samples and assign theta values for exon set negative binomial distributions for each sample.

**Note:** The theta range here is not used since there is no longer a negative binomial component to the simulation. 

```{r}
# Output: list of probability matrix (p_mat), count matrix (c_mat), gene names(gene_names),
#     and theta parameters for n*J samples. ct_mat will be modified in diff_exp() later
#     p_mat and c_mat: first n columns = CT1, second n cols CT2, third n cols CT3; rows = all genes
geneInfo = gene_level2(total_cts = total_cts, gene_alpha = gene_alpha, 
                       theta_range = 10:20, seed = seeds[2])

```

```{r, eval=FALSE}
# save(geneInfo, file = "Simulated_Output/initial_geneInfo2.RData")
```

Select genes for differential expression (100 of 1000) and differential isoform usage (100 of 1000, not overlapping with differential expression genes).

```{r}
# Output: matrix where col 1 ("diffExp") = boolean value of whether gene is differentially 
#     expressed (1) or not (0), and col 2 ("diffUsg") = boolean value of whether gene will have
#     differential isoform usage (1) or not (0). Rownames correspond to gene names
diffGeneInfo = diff_genes(CT1_counts = geneInfo$ct_mat[,1:n], nTE_filtered = nTE_filtered, 
                          num_diff = 200, seed = seeds[3])

# Check that chosen genes for diff exp and diff iso usage came from nTE_filtered genes
sum(rownames(diffGeneInfo[which(diffGeneInfo[,"diffExp"] == 1),]) %in% nTE_filtered$geneId)
sum(rownames(diffGeneInfo[which(diffGeneInfo[,"diffUsg"] == 1),]) %in% nTE_filtered$geneId)
sum(rownames(diffGeneInfo[which(diffGeneInfo[,"noDiff"] == 1),]) %in% nTE_filtered$geneId)
```

```{r}
# Identify and save gene names specified for differential expression and differential isoform usage
# differential expression
genesDiffExp = rownames(diffGeneInfo[which(diffGeneInfo[,"diffExp"] == 1),])
# differential isoform usage
genesDiffUsage = rownames(diffGeneInfo[which(diffGeneInfo[,"diffUsg"] == 1),])
# genes with no differential expression/usage to be used for false positives
genesNoDiff = rownames(diffGeneInfo[which(diffGeneInfo[,"noDiff"] == 1),])
# save results in data.frame
genes_df = data.frame(diffExp = genesDiffExp, diffUsg = genesDiffUsage, noDiff = genesNoDiff)

```

```{r, eval=FALSE}
# Same genes_df as in original simulation (Sim1)
# save(genes_df, file = "Simulated_Output/genes_simulated_w_diffExp_diffUsg.RData")
```

Specify and apply fold changes to gene counts for genes specified to have differential expression in one of the cell types (Cell Type 2). A random component was introduced to the fold change so there is some variation across samples.

```{r}
# Output: New gene counts matrix where CT2 has differentially expressed genes compared to CT1 and CT3
#     matrix same dimension and organization of geneInfo$ct_mat
geneCtDiffExp = diff_exp(gene_counts = geneInfo$ct_mat, n = n, J = J, 
                         CT_diffExp = 2, diff_genes_mat = diffGeneInfo, 
                         propUp = 0.30, seed = seeds[4])
```


## Pure Sample Set-Up - Sim 3 Fit 3 (Blueprint Alpha from 11A Series Used)

Find cluster, isoform, and exon set level information for the 1,000 genes of interest for all cell types. 

For the 100 genes specified to have differential isoform usage: each cell type will have a different dirichlet distribution specified for these isoforms across the samples of the different cell types. 

Description of different dirichlet distribution used:

In the Blueprint simulation, alpha values for each cluster were saved during the IsoDeconvMM algorithm fit. Using the cluster results from the `isoDetector` function in the `isoform` R package, we identified clusters that had at least one isoform with high abundance in one cell type but low or non-existent abundance in the other two cell types.

In the document "Blueprint Alpha Parameters.Rmd", alpha values from a Blueprint simulation (simulation 11A) were saved for clusters that had between 3 and 15 isoforms. These values will be used as hyperparameters in the isoform Dirichlet distributions to make these distributions more realistic.

The above mentioned Blueprint alpha parameters will be used for those genes specified to have differential isoform usage. For all other genes, they will be given a generic uniform isoform distribution (for all three cell types, all alpha values describing the isoform Dirichlet distribution will have the value of 10).

The gene counts for the 'other' 4,172 genes will be approx. evenly distributed among the singular exon sets associated with the gene cluster. 

```{r}
alpha_collect = alpha_collect11A

# Determine which genes should have differential isoform usate
# Note: gene clusters chosen so number isoforms 3 <= I <= 15 for all genes
diffUsg = ifelse(nTE_filtered$geneId %in% genesDiffUsage, 1, 0)

names(diffUsg) = nTE_filtered$geneId

# Check
sum(diffUsg)

# Alpha collection to use for each gene
## Will store the alpha parameters to use for all three cell types
## Selection of correct cell type parameters will occure in iso_exon_info_MM() function
alpha_record = list()

# names of 1,000 clusters of interest used in simulation
clust_names = nTE_filtered$clustID
 
set.seed(1323)
for(clust in clust_names){
    
    # name of gene associated with cluster
    gene = nTE_filtered$geneId[which(clust_names == clust)]
    # Effective length matrix - ExI (num exon sets x num isoforms)
    X = X_info[[clust]]$X
    # number isoforms
    I = ncol(X)
    # Indicator to whether this gene should have differential isoform usage
    indic = diffUsg[which(names(diffUsg) == gene)]
    if(indic == 1){
      # dirichlet alpha parameters for isoforms
      ## Randomly select a saved alpha parameter collection from saved Blueprint results
      ## based on number of isoforms (dimension Ix3)
      samp = sample(1:length(alpha_collect[[I]]), size = 1)
      alpha = alpha_collect[[I]][[samp]]
      
    }else{
      # Give isoforms arbitrary uniform distribution (same for all cell types)
      alpha = matrix(10, nrow = I, ncol = 3)
    }
    
    alpha_record[[clust]] = alpha
}

save(alpha_record, file = "Simulated_Output/alpha_record_sim3fit3.RData")

# Output for iso_exon_info(): for each cluster corresponding to the 1,000 genes of interest, 
#     records the alpha parameters for the isoform dirichlet dist (iso_alpha), 
#     the probability vector associated with the isoform distribution (rho),
#     the gamma values used in the calculating of the multinomial probabilities (gamma_vals), 
#     the multinomial probabilities (mm_probs),
#     the counts for each gene/cluster (gene_ct),
#     and a character vector of the relevant exon sets (exon_sets)

iso_exon_1000_3C = list()

iso_exon_1000_3C[["CT1"]] = iso_exon_info_MM(genes_info = geneCtDiffExp[which(rownames(geneCtDiffExp) %in% nTE_filtered$geneId),1:n],
                                       nTE_filtered = nTE_filtered, EffLen_info = X_info, 
                                       CT = 1, alpha_record = alpha_record, seed = seeds[5])

iso_exon_1000_3C[["CT2"]] = iso_exon_info_MM(genes_info = geneCtDiffExp[which(rownames(geneCtDiffExp) %in% nTE_filtered$geneId),(n+1):(2*n)],
                                       nTE_filtered = nTE_filtered, EffLen_info = X_info, 
                                       CT = 2, alpha_record = alpha_record, seed = seeds[6])

iso_exon_1000_3C[["CT3"]] = iso_exon_info_MM(genes_info = geneCtDiffExp[which(rownames(geneCtDiffExp) %in% nTE_filtered$geneId),(2*n+1):(3*n)],
                                       nTE_filtered = nTE_filtered, EffLen_info = X_info, 
                                       CT = 3, alpha_record = alpha_record, seed = seeds[7])

```

```{r, eval=FALSE}
save(iso_exon_1000_3C, file = "Simulated_Output/Iso_Exon_Params_1000_Sim3Fit3.RData")
```

## 'Other' Exon Sets

```{r}

# Identify 'other' singular exon sets
exon_sets_other = other_geneInfo$exon_sets_other
nTE_other = other_geneInfo$nTE_other

# Output for other_exonset_count(): for each cluster corresponding to the 'other' 4,017 genes,
#     records the exon sets with single exons for all isoforms (exon_sets),
#     and a counts matrix for each of these singular exon sets based on arbitrarily 
#     distributing the count for a gene approx. uniformly across these exon sets.
iso_exon_other = list()

iso_exon_other[["CT1"]] = other_exonset_count(genes_info = geneCtDiffExp[which(rownames(geneCtDiffExp) %in% nTE_other$geneId),1:n],
                                              nTE_other = nTE_other,
                                              exon_sets_other = exon_sets_other,
                                              iso_dist = rep("uniform", times = nrow(nTE_other)),
                                              alphaRange = c(20,50), seed = seeds[8])

iso_exon_other[["CT2"]] = other_exonset_count(genes_info = geneCtDiffExp[which(rownames(geneCtDiffExp) %in% nTE_other$geneId),(n+1):(2*n)],
                                              nTE_other = nTE_other,
                                              exon_sets_other = exon_sets_other,
                                              iso_dist = rep("uniform", times = nrow(nTE_other)),
                                              alphaRange = c(20,50), seed = seeds[9])

iso_exon_other[["CT3"]] = other_exonset_count(genes_info = geneCtDiffExp[which(rownames(geneCtDiffExp) %in% nTE_other$geneId),(2*n+1):(3*n)],
                                              nTE_other = nTE_other,
                                              exon_sets_other = exon_sets_other,
                                              iso_dist = rep("uniform", times = nrow(nTE_other)),
                                              alphaRange = c(20,50), seed = seeds[10])
```

```{r, eval=FALSE}
# save(iso_exon_other, file = "Simulated_Output/Iso_Exon_Counts_Other.RData")
```

## Record Exon Set Counts in counts.txt Files

In this step, exon set counts are (a) simulated from the exon set negative binomial parameters for the 1,000 genes of interest and (b) recorded from the `iso_exon_other` list output. These exon set counts are then concatenated into one file for each sample and stored in a counts.txt file in the same output format as the function `countReads()` from the `isoform` R package. These count files are stored in the "Simulated_Output/Pure Sample Counts" folder.

```{r}

library(stringr)

# Simulate and record exon set counts 
## count.txt files saved in sub-folder of Simulated_Output

rep_pure = c(str_c("0",1:9),10:n)

files1 = str_c("CT1_ref_",rep_pure,"_counts")
files2 = str_c("CT2_ref_",rep_pure,"_counts")
files3 = str_c("CT3_ref_",rep_pure,"_counts")

```

```{r, eval=FALSE}

# Using 11A series alpha values
counts_output_MM(exonInfo_1000 = iso_exon_1000_3C, exonInfo_other = iso_exon_other,
                n = 20, file_labels = c(files1, files2, files3),
                folder = "Simulated_Output/Sim3_Fit3_Pure_Sample_Counts", seed = seeds[11])

```


## Create Mixture Samples

The mixture files will be created during the simulation process. They will be created using the `mixture_creation2` function in the file "sim_functions.R".  

See "sim3_fit3_mixtures.R" code in "sim3_code" folder.


## Create Fragment length files for mixtures samples

For the purposes of this simulation, we will assume that the fragment length distribution is the same across all samples. Therefore, in an attempt to simplify the number of files created and used for the simulation, only one fragment length distribution file will be created. This file will be recycled for all `IsoDeconvMM` fit procedures for all mixture samples.

```{r, eval=FALSE}

fragLens_out = function(total_reads = 2*10^7, mean = 300, SD = 50, lenMin = 150, lenMax = 600,
                        out_file, seed){
  # fragLens_dist() in geneModel_code/fragLens_dist.cpp file
  freq_dist = fragLens_dist(total_reads, mean, SD, lenMin, lenMax)
  freq_dist = freq_dist[which(freq_dist[,1] > 0),]
  write.table(freq_dist, file = out_file, col.names = F, row.names = F)
}

fragLens_out(total_reads = 2*10^7, mean = 300, SD = 50, lenMin = 150, lenMax = 600,
             out_file = "Simulated_Output/sim_fraglens.txt", seed = seeds[14])

```

The End